win_psexec: execute cmds on remote systems as any user

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_psexec

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
The psexec tool is quite popular among Windows admins for automating tasks between systems. This module makes the tool available so that you can connect to remote servers with elevated privileges.

I wasn't sure if we want to have this module. If this is acceptable, I will also add the documentation + examples.

```yaml
- win_psexec:
    command: whoami.exe
    hostnames: [ remote_system ]
    username: DOMAIN\user
    password: password
    chdir: C:\Windows\Temp\
    priority: high
```

dagwieers committed:
win_psexec: execute cmds on remote systems as any user

This module uses the versatile psexec tool to run any command remotely
as any user (incl. domain users).

dagwieers commented:
I am adding documentation as we speak, and am going to reformat the code based on the new stuff I learned yesterday from @nitzmahone !

dagwieers committed:
Add missing documentation

Now that this module is deemed acceptable for inclusion,
the documentation is an essential part.

dagwieers committed:
win_psexec: Small cosmetic changes

dagwieers committed:
win_psexec: add more options (priority, elevated, ...)

dagwieers committed:
Fixes after more testing

dagwieers commented:
Ha, documentation tests suddenly work !
No change to the documentation though.

jhawkesworth commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,142 @@
+'''
+
+RETURN = '''
+cmd:
```
Any reason for calling this 'cmd' in the return but the corresponding module param is 'command'?  I'd be tempted to make them the same, probably going with 'command', just because whole words communicate better than the terse, vowel-free versions we are all so used to from the command line.

dagwieers replied:
Well, it does not have the same content as `command`, so I did not want to give it the same name.

- command: is the original command you want to run remotely from psexec
- cmd: is the complete command line run on the target, including the psexec command line

I am open to change it, I was also contemplating using `command_line` instead. But since we have no standard naming (which I think is a shame) we can do whatever we like here.

jhawkesworth replied:
That's a perfectly good reason to have a different name.

Standard naming would be good, although hard to enforce and could compromise readability (which is something I value highly in ansible).

command_line makes sense to me but can live with either.

dagwieers replied:
Maybe psexec_command is more descriptive in this context ?

dagwieers replied:
The only reason for me to output this was to be sure win_psexec was doing the right thing. And I can imagine some people will need the same thing if they have issues using it (wrt. quoting etc.). So that you can copy&paste and do exactly what the module did.

nitzmahone replied:
+1 to `psexec_command` just for clarity...

dagwieers committed:
Renamed 'cmd' to 'psexec_command' + more

- Also replaced PSObject() with a hash table
- Made $chdir of type "path"
- Renamed $args to $extra_args

dagwieers commented:
ready_for_review

dagwieers committed:
Various improvements

- Switched to using booleans for most parameters
- Added type 'bool' to boolean parameters
- Added 'interactive' parameter
- Added 'wait' parameter
- Added an interactive example

dagwieers commented:
ready_for_review

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+# WANT_JSON
+# POWERSHELL_COMMON
+
+# See also: https://technet.microsoft.com/en-us/sysinternals/pxexec.aspx
```
was going to say this URL looks wrong, but it really is pxexec.aspx. Weird...

dagwieers replied:
Yup, I wanted to fix it once as well ;-)

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+
+# See also: https://technet.microsoft.com/en-us/sysinternals/pxexec.aspx
+
+$params = Parse-Args $args;
```
What, no check_mode support? Wimp. ;)

dagwieers replied:
Haha, didn't think of it honestly. Will be easy to add :-P

dagwieers replied:
Maybe add creates/removes though ?

nitzmahone replied:
Probably not- would only make sense for "local" run, right (because you're usually running on a 3rd machine).

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+
+$params = Parse-Args $args;
+
+$command = Get-AnsibleParam -obj $params -name "command" -default "whoami.exe"
```
Why default this to whoami instead of just making required + no default?

dagwieers replied:
I got the idea from win_reboot :-D
I thought it would be nice because whoami would tell you immediately if it worked.
But sure `whoami' is a weird default outside from testing. Will fix !

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+$wait = Get-AnsibleParam -obj $params -name "wait" -type "bool" -default $true
+$noprofile = Get-AnsibleParam -obj $params -name "noprofile" -type "bool" -default $false
+$elevated = Get-AnsibleParam -obj $params -name "elevated" -type "bool" -default $false
+$limited = Get-AnsibleParam -obj $params -name "limited" -type "bool" -default false
```
s/false/$false/

dagwieers replied:
Ouch.

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+$password = Get-AnsibleParam -obj $params -name "password"
+$chdir = Get-AnsibleParam -obj $params -name "chdir" -type "path"
+$wait = Get-AnsibleParam -obj $params -name "wait" -type "bool" -default $true
+$noprofile = Get-AnsibleParam -obj $params -name "noprofile" -type "bool" -default $false
```
Are these `-type "bool"` from a hacked powershell.ps1, or just no-op? Wouldn't be opposed to lighting that up (again, will be superseded by declarative argspec, but still...)

dagwieers replied:
Right, I wanted to add bool-support, but wasn't sure if you liked that. I noticed some modules doing:

```powershell
get-Attr $params "parameter" $true | ConvertTo-Bool
```

So the idea was to do the same for -type "bool", nothing more, nothing less.

nitzmahone replied:
Yeah, I hate that pattern (because it masks other issues), so if you want to hack up `-type bool` support for powershell.ps1, I'd merge it.

dagwieers replied:
Done !

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+}
+
+If ($chdir -ne $null) {
+    $arguments += " -w $chdir"
```
$chdir should probably be quoted (eg, C:\Program Files\blah)

dagwieers replied:
Well, to be honest, I did that for the username and password and psexec tripped over it.
So I figured psexec must be smart in parsing, but I did not test. Will fix, if needed !

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+$proc.Start() | Out-Null
+$result.stdout = $proc.StandardOutput.ReadToEnd()
+$result.stderr = $proc.StandardError.ReadToEnd()
+$proc.WaitForExit()
```
Reading the outputs sequentially like this will deadlock psexec itself on large stderr outputs (> ~4k natively)- see win_shell/win_command for a way to deal with this (unfortunately harder in raw PS)- you basically need to either use event-driven reads or use two threads to keep both buffers empty.

dagwieers replied:
Ok. I haven't had any issues myself, but maybe I was lucky.

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,130 @@
+    $result.failed = $true
+}
+
+Exit-Json $result
```
Needs trailing newline

dagwieers replied:
Still don't understand exactly why that is. VIM is eating those away for breakfast.

dagwieers replied:
Done !

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+    description:
+    - The command line to run through PsExec (limited to 260 characters).
+    required: true
+    default: whoami.exe
```
don't forget to remove if you change default

dagwieers replied:
Ah, weird that I had `required: true` there anyway. Fixed !

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+
+EXAMPLES = r'''
+# Test the PsExec connection to the local system (target node) with your user (runs whoami)
+- win_psexec:
```
ah, now I see why, but still, would prefer `command` to be required.

dagwieers replied:
Bingo ! :-)

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+    description: The complete command line used by the module, including PsExec call and additional options.
+    returned: always
+    type: string
+    sample: 'psexec.exe \\remote_server -u DOMAIN\Administrator -p some_password E:\setup.exe'
```
Does this one need to be quoted? (not sure, but don't think so)

dagwieers replied:
In the old days, the YAML parser tripped over a second colon.
So I still have this weird habit of quoting whatever string with a colon by default.
I only noticed recently that's not needed anymore (with e.g. Windows paths).

Will fix.

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+    description: The standard output from the command
+    returned: always
+    type: string
+    sample: 'Success.'
```
unnecessary quotes?

dagwieers replied:
Will fix.

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+    description: The error output from the command
+    returned: always
+    type: string
+    sample: 'Error 15 running E:\setup.exe'
```
unnecessary quotes?

dagwieers replied:
Will fix.

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+    description: Possible error message on failure
+    returned: failed
+    type: string
+    sample: "The 'password' parameter is a required parameter."
```
unnecessary quotes?

dagwieers replied:
Will fix.

nitzmahone reviewed (changes requested):
A few minor things- the big one is the deadlock on large error text.

dagwieers committed:
Added -type "bool" support to Get-AnsibleParam

dagwieers committed:
Fix deadlock

nitzmahone commented on lib/ansible/modules/windows/win_psexec.py:
```diff
@@ -0,0 +1,159 @@
+module: win_psexec
+version_added: '2.3'
+short_description: Runs commands (remotely) as another (privileged) user
+description:
```
should probably have `requires: [ psexec ]`

dagwieers replied:
Done.

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,180 @@
+    }
+}
+'@
+
```
Should probably have a `Get-Command $executable` with a message that psexec is required to fail gracefully if it's not there.

dagwieers replied:
Done.

dagwieers commented:
ready_for_review

nitzmahone commented on lib/ansible/modules/windows/win_psexec.ps1:
```diff
@@ -0,0 +1,189 @@
+    changed = $true
+}
+
+If (-Not (Get-Command $executable)) {
```
I think you need -ErrorAction SilentlyContinue on this one or it bombs under Strict + Stop.

dagwieers replied:
Hmm, I tested it and it worked. Proper error with Fail-Json

dagwieers commented:
I also tested `-type "bool"` and it works as expected.

dagwieers committed:
When using `wait:no` return code is PID of process
